How Should I Style My React Application? (#227)
leerob/leerob.io: pages/blog/css-with-react.mdx

---
title: 'How Should I Style My React Application?'
publishedAt: '2020-09-15'
summary: 'TODO'
image: '/static/images/css-with-react/banner.png'
---

> The React ecosystem is unopinionated. **This is both it's greatest strength and weakness.**

Reviewer: I am not a native speaker but I think that the ecosystem can be opinionated – React however is not a full blown framework and therefore doesn't prescribe how to do many things including styling.

Author: That's a good clarification! Let me tweak this.

Over the years, I tried about seven different ways to style my React applications. Each solution has their own pros and cons.

> Most comparions between CSS / CSS-in-JS and their ecosystem fail to consider the scale of the application and the size of the team.

Reviewer: ```suggestion
Most comparisons between CSS / CSS-in-JS and their ecosystem fail to consider the scale of the application and the size of the team.
```

Let's go on a journey and learn how CSS has evolved to bring us to its current state. I'll share my journey and then discuss pros / cons of
different approaches for styling React applications after. I'll try to make this comparison as unopinionated as possible.

## Vanilla CSS

Like most, I started web development with vanilla CSS. You have to learn the basics first. All we need is an HTML and CSS file.

```html:index.html
<h1 class="header">Welcome</h1>
```

```css:index.css
.header {
  font-size: 32px;
}
```

As I built more applications, I began to understand some of the drawbacks.

- **Problem.** todo

> - **Specificity.** Since CSS has a global namespace, you can unintentionally target too many elements.

Reviewer: ```suggestion
- **Global Namespace.** Since CSS has a global namespace, you can unintentionally target too many elements.
```

Probably global namespace is more accurate. Specificity is the weight of a selector

- **Colocation. **It's hard to modularize your CSS, which makes it difficult to delete code safely.

At this point, I was introduced to CSS pre-processors.

> ## Saas

Reviewer: ```suggestion
## Sass
```

> Saas, one of the most popular, allowed you to write more reusable, maintainable CSS. Some of the most popular features are variables, mixins, and modules.

Reviewer: ```suggestion
Sass, one of the most popular preprocessors, allows you to write more reusable, maintainable CSS. Some of the most popular features are variables, mixins, and modules.
```

- **Variables.** I want to define global values once and share them throughout my stylesheets.
- **Mixins.** I want to reuse particular snippets of CSS.
- **Modules.** I want to split up my CSS files for a more maintainable codebase.

```scss:_base.scss
$font-stack: Helvetica, sans-serif;
$primary-color: #333;

body {
  font: 100% $font-stack;
  color: $primary-color;
}
```

```scss:_header.scss
@use 'base';

.header {
  font-size: 32px;
}
```

Sass helped me write more modular, reusable CSS – at least, for a while. Over time, I ran into new issues.

- **Naming.** As your CSS grows, it's easy to have class names get out of hand.
- **Browser support.** Certain features aren't supported across all browsers and require "vendor prefixes" like `-moz` (Firefox), `-webkit` (Safari), and `-ms` (Internet Explorer).

> To solve naming, I started to adopt [BEM](http://getbem.com/introduction/). Block Element Modifier is a methodology that helps you to create reusable components and code sharing in front-end development.

Reviewer: organize css in a modular way and keep specificity low

## Pros and Cons

### Vanilla CSS

import ProsCard from '../../components/ProsCard';
import ConsCard from '../../components/ConsCard';

<ProsCard
  title="vanilla CSS"
  pros={['No toolchain necessary. Works with every browser & language.']}
/>

<ConsCard
  title="vanilla CSS"
  cons={['Not a viable option for large web applications.']}
/>

### CSS Modules

[CSS Modules](https://github.com/css-modules/css-modules) would be my choice for smaller teams that don't need to share
components across applications.

[Next.js](https://nextjs.org) has support for CSS Modules, which means you don't have to worry about setting up [Webpack]().
It also sets up [autoprefixer](https://css-tricks.com/autoprefixer/), which makes CSS Modules an attractive alternative to CSS-in-JS in some cases.

<ProsCard
  title="CSS modules"
  pros={[
    'Class names are scoped to the corresponding component.',
    'CSS variables allow you to reuse values.',
    'Your application can work without JavaScript.',
    'Lowest barrier to entry.',
    'Works with any language.'
  ]}
/>

<ConsCard
  title="CSS modules"
  cons={[
    'Requires confguring a toolchain like Webpack.',
    'Distribution is hard.',

>     'Lack of nesting and CSS linting.',

Reviewer: these can be done with additional tooling eg postcss/sass and stylelint

    `Some argue there's more context switching between different files`
  ]}
/>

### CSS-in-JS

There's many different CSS-in-JS libraries. The two most popular are [styled-components](https://styled-components.com/) and [Emotion](https://emotion.sh/docs/introduction).

There's also zero-runtime solutions like [Linaria](https://github.com/callstack/linaria), where CSS is extracted to CSS files at build-time.
To see comparisions between different libraries, see [CSS-in-JS benchmarks.](https://github.com/A-gambit/CSS-IN-JS-Benchmarks)

<ProsCard
  title="CSS-in-JS"
  pros={[
    'Easy to distribute as reusable components on NPM.',
    'You can take advatange of anything in the JavaScript ecosystem.',
    'Styles are colocated with components, making it easy to delete code.',
    'Easy to provide multiple themes.'
  ]}
/>

<ConsCard
  title="CSS-in-JS"

>   cons={[

Reviewer: I'd add naming is hard here too since you need to name those component :P Also they use context extensively (in contrast css + custom properties are much faster)

    'All libraries that follow System UI spec should be interoperable.'
  ]}
/>

<ConsCard
  title="Theme UI / styled-system"
  cons={[
    'You dislike providing styles as props.',
    'Some argue it breaks separation of concerns.'
  ]}
/>

### Tailwind CSS

[Tailwind CSS](https://tailwindcss.com/) has been rapidly growing in popularity. Many people are fed up with the bloat of utility frameworks like
[Bootstrap](https://getbootstrap.com/) and want an alternative. Tailwind provides an impressive developer experience with a refined API.

With the combination of [Tailwind UI](https://tailwindui.com/), you can easily build extensible, well-designed applications.

<ProsCard
  title="Tailwind"
  pros={[
    'Utility classes allow you to never leave your HTML.',
    'Allows you to easily enforce a design system.',
    'Spend less time trying to name components',
    'All libraries that follow System UI spec should be interoperable.'
  ]}
/>

<ConsCard
  title="Tailwind"
  cons={[
    'You can have really, really long class names.',
    'To use effectively, requires toolchain with PurgeCSS.',
    'You need to learn Tailwind syntax, which is slighty different than standard CSS.'
  ]}
/>

> 

Reviewer: at the end I would add a poll or ask for comments on Twitter (can bump engagement)

- Your team
- The size of your application
- What you're trying to build

## Pros and Cons

### Vanilla CSS

import ProsCard from '../../components/ProsCard';
import ConsCard from '../../components/ConsCard';

<ProsCard
  title="vanilla CSS"
  pros={['No toolchain necessary. Works with every browser & language.']}
/>

<ConsCard
  title="vanilla CSS"
  cons={['Not a viable option for large web applications.']}
/>

### CSS Modules

[CSS Modules](https://github.com/css-modules/css-modules) would be my choice for smaller teams that don't need to share
components across applications.

[Next.js](https://nextjs.org) has support for CSS Modules, which means you don't have to worry about setting up [Webpack](https://webpack.js.org/).
It also sets up [autoprefixer](https://css-tricks.com/autoprefixer/) making CSS Modules an attractive alternative to CSS-in-JS in some cases.

<ProsCard
  title="CSS modules"
  pros={[
    'Class names are scoped to the corresponding component.',
    'CSS variables allow you to reuse values.',
    'Your application can work without JavaScript.',
    'Lowest barrier to entry.',
    'Works with any language.'
  ]}
/>

<ConsCard
  title="CSS modules"
  cons={[

>     'Requires confguring a toolchain like Webpack.',

Reviewer 2: ```suggestion
    'Requires configuring a toolchain like Webpack.',
```

Author: Thank you! 💯
